mitmproxy/protocol/socks_client.py

```diff
@@ -0,0 +1,86 @@
+import socket
+import six
+import sys
+
+from netlib.exceptions import TcpException
+from netlib import (tcp, socks)
+from netlib.socks import (
+    METHOD, VERSION, CMD, USERNAME_PASSWORD_VERSION, REP,
+    UsernamePasswordAuth, UsernamePasswordAuthResponse,
+    ClientGreeting, ServerGreeting, Message, SocksError
+)
+
+from .base import (Layer, ProxyServerConnection)
+from ..exceptions import ProtocolException
+
+class SocksClientLayer(Layer):
+
+    def __init__(self, ctx, server_tls):
+        super(SocksClientLayer, self).__init__(ctx)
+        self.server_conn = ProxyServerConnection(
+            None,
+            self.ctx
+        )
+        self.server_tls = server_tls
+        self.socks_address = tcp.Address((u"127.0.0.1", 1080))
+        self.ctx.set_server(self.socks_address)
+        self.socks_username = "test"
+        self.socks_password = "test"
+
+    def __call__(self):
+        layer = self.ctx.next_layer(self)
+        layer()
+
+    def _socks_username_password_auth(self):
+        rfile, wfile = self.server_conn.rfile, self.server_conn.wfile
+        cauth = UsernamePasswordAuth(USERNAME_PASSWORD_VERSION.DEFAULT, self.socks_username, self.socks_password)
+        cauth.to_file(wfile)
+        wfile.flush()
+        sauth = UsernamePasswordAuthResponse.from_file(rfile)
+        if sauth.status != 0:
+            raise SocksError(0, "authentication failed")
+
+    def change_socks_proxy_server(self, address, username = None, password = None):
```

Comment:
This is a great idea - how about we call it `change_proxy_server` for both `http` and `socks`?